### PR TITLE
Game: Refactor Game Simulator

### PR DESCRIPTION
**Pull request by lhw-1**
### Description

Partially addresses #2806. Refactors the Game Simulator code to improve code quality & extensibility.

While changes are mostly on code quality, there are minor changes to UI:
- Asset Uploader is now renamed to Asset Viewer.
  - A new [tab](https://blueprintjs.com/docs/#core/components/tabs) is introduced to switch between asset viewer and the asset uploader.
- Checkpoint Simulator is now renamed to Chapter Simulator.
- Chapter Simulator (the name of which was potentially confusing) is now rightly renamed to Chapter Publisher.
  - The Chapter Simulation option is no longer available for Chapter Publisher - it is planned to be re-introduced once the next stage of the refactoring is complete.

There are two more major updates planned following this refactor:
- Incorporate lazy loading of the game within the game simulator itself - the game is not essential for the entire game simulator, though it is necessary for chapter simulation, meaning that it should be lazily loaded.
- Change panel to a popup - this will be part of the UI update for the Game Simulator.

Documentation update & adding tests will be done after the above two updates.

### Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality) - _The tabs have been newly introduced, though minor._
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update
- [X] Code quality improvements

### Checklist

- [X] I have tested this code

### CODE REVIEW

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+import { ChapterDetail } from 'src/features/gameSimulator/GameSimulatorTypes';
+
+import ChapterPublisherEditor from './ChapterPublisherEditor';
+
+/**
+ * This components renders the Chapter Publisher component in the Game Simulator.
+ *
+ * @param textAssets - List of all text assets on S3 to choose from.
+ */
+const ChapterPublisher = memo(() => {
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx:14]: Nit: Can we, to keep it consistent with the other codebases:

```tsx
import React, { memo, useState } from 'react';
// ...
const ChapterPublisher: React.FC = () => { /* ... */ };

// ...
export default memo(ChapterPublisher, () => true);
```

Also, can I check, should the component never be recreated after it's been created once? If yes, I'd say `() => true` is preferred in the memoization condition.

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  const { value: textAssets } = useRequest<string[]>(fetchTextAssets, []);
+  const { value: chapters } = useRequest<ChapterDetail[]>(fetchChapters, []);
+
+  const [chosenIndex, setChosenIndex] = useState(-1);
+
+  return (
+    <>
+      <h3>Publish / Edit Chapters</h3>
+      <select className="bp4-menu" onChange={(e: any) => setChosenIndex(e.target.key)}>
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx:23]: We should be using `bp5-menu` since the CSS namespace is now Blueprint 5.

File: src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewerUpload.tsx [hunk 3]
```diff
@@ -51,32 +46,28 @@ const AssetFileUploader = () => {
...
+      <Popover placement={Position.BOTTOM} content={uploadButtonPopoverContent}>
         <Button text={uploadFolder} />
-      </Popover2>
+      </Popover>
+      <br />
+      <br />
       {showfolderOverwrite && (
         <InputGroup
-          placeholder={folderOverwritePlaceholder}
+          placeholder={"Specify your own custom folder: e.g. 'locations/hallway'"}
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewerUpload.tsx:62]: Nit: Unnecessary curly braces around string.

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        {chapters.map((chapter, chapterIndex) => (
+          <option key={chapterIndex} value={chapter.title}>
+            {`Chapter ${chapterIndex}: ${chapter.title}`}
+          </option>
+        ))}
+        <option key={defaultChapter.id} value={''}>{`New chapter`}</option>
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx:29]: Nit: Unnecessary curly braces, template literal

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx [hunk 5]
```diff
@@ -81,74 +71,79 @@ const ChapterEditor = React.memo(({ chapterDetail, checkpointFilenames }: Chapte
...
       </h4>
       <b>Open date: </b>
       {openDate && getStandardDateTime(openDate.toISOString())}
       <DatePicker
-        onChange={(date: Date) => {
+        onChange={(date: Date | null) => {
           date && setOpenDate(date);
         }}
+        showActionsBar={true}
+        highlightCurrentDay={true}
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx:127]: Nit: Unnecessary `={true}`

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx [hunk 6]
```diff
@@ -160,17 +155,14 @@ const ChapterEditor = React.memo(({ chapterDetail, checkpointFilenames }: Chapte
...
-      <br />
+      <hr />
       <br />
       <Switch
         checked={isPublished}
         labelElement={'Published'}
         onChange={() => setIsPublished(!isPublished)}
       />
-      <Button onClick={saveChapter}>Save Changes</Button>
+      <Button onClick={saveChapter}>Save Changes</Button>{' '}
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx:165]: Was this change intentional?

File: src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulator.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import ChapterSimulatorTextLoader from './ChapterSimulatorTextLoader';
+
+/**
+ * This component renders the Chapter Simulator component in the Game Simulator.
+ *
+ * It will simulate a game chapter using the given text files (from either S3 or Local).
+ *
+ * @param textAssets List of all text assets on S3 to choose from.
+ */
+const ChapterSimulator = () => {
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulator.tsx:17]: Ditto on typing with `React.FC`

File: src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulatorTextLoader.tsx [hunk 8]
```diff
@@ -1,25 +1,23 @@
...
+ * This component allows chapter text files to be loaded either from S3, or from the user's local device.
  *
- * @param storageName the field in browser storage where the loaded/fetched txt files get stored temporarily
- * @param s3TxtFiles the list of S3 txt files to choose from
+ * @param storageName The field within browser storage to temporarily store the loaded / fetched text file(s).
+ * @param s3TxtFiles List of all text assets on S3 to choose from.
  */
-function CheckpointTxtLoader({ storageName, s3TxtFiles }: Props) {
-  const [chosenFilename, setChosenFilename] = React.useState(s3TxtFiles[0]);
+function ChapterSimulatorTextLoader({ storageName, s3TxtFiles }: StorageProps) {
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulatorTextLoader.tsx:19]: Ditto on typing, also perhaps we can use lambda functions for code style consistency (nit).

File: src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+import AssetViewerUpload from './AssetViewerUpload';
+import { convertAssetPathsToTree, treeMap } from './AssetViewerUtils';
+
+/**
+ * This component renders the Asset Viewer component in the Game Simulator.
+ *
+ * It provides a preview of all the S3 asset files in a document tree format.
+ * The selected asset will be available for preview.
+ */
+const AssetViewer = () => {
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx:18]: Ditto on `React.FC`

File: src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+  const { value: assetPaths } = useRequest<string[]>(fetchAssetPaths, []);
+
+  const [currentAsset, setCurrentAsset] = React.useState('');
+  const [assetTree, setAssetTree] = React.useState([] as TreeNodeInfo[]);
```
Comment (RichDom2185) [inline src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx:22]: Nit: We should prefer passing a generic type argument instead of a typecast.

```suggestion
  const [assetTree, setAssetTree] = React.useState<TreeNodeInfo[]>([]);
```

**Review by RichDom2185 — CHANGES_REQUESTED** [review]
Hi, thanks for this, looking forward to the refactor! I just have some comments below, let me know your thoughts!

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx (hunk 1 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx]: Addressed in [962b341](https://github.com/source-academy/frontend/pull/2836/commits/962b3413e413ff74052dd0e7f90ff5830ec709fb) - and to answer the second part, `() => true` is applicable for this component, yes 👍

File: src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulator.tsx (hunk 7 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulator.tsx]: Addressed in [962b341](https://github.com/source-academy/frontend/pull/2836/commits/962b3413e413ff74052dd0e7f90ff5830ec709fb).

File: src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx (hunk 9 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx]: Addressed in [962b341](https://github.com/source-academy/frontend/pull/2836/commits/962b3413e413ff74052dd0e7f90ff5830ec709fb).

File: src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulatorTextLoader.tsx (hunk 8 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterSimulator/ChapterSimulatorTextLoader.tsx]: Addressed in [962b341](https://github.com/source-academy/frontend/pull/2836/commits/962b3413e413ff74052dd0e7f90ff5830ec709fb). Thanks for the catch! Forgot to convert to lambda 😅

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx (hunk 2 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx]: Addressed in [d6b805d](https://github.com/source-academy/frontend/pull/2836/commits/d6b805d1a82065dea06d44dae8eff4438b1abcf7).

File: src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewerUpload.tsx (hunk 3 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewerUpload.tsx]: Addressed in [d6b805d](https://github.com/source-academy/frontend/pull/2836/commits/d6b805d1a82065dea06d44dae8eff4438b1abcf7).

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx (hunk 4 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisher.tsx]: Addressed in [d6b805d](https://github.com/source-academy/frontend/pull/2836/commits/d6b805d1a82065dea06d44dae8eff4438b1abcf7).

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx (hunk 6 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx:165]: The change was intentional to have some whitespace between two buttons - though on hindsight I'll revert this for now and look at it a bit more systematically in the UI update.

Do we prefer directly changing margins over using `{' '}` for quick whitespacing?

File: src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx (hunk 10 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/assetViewer/AssetViewer.tsx]: Addressed in [664d352](https://github.com/source-academy/frontend/pull/2836/commits/664d3528c4fc83f110bb07756f944cd7778ae46e). Thanks for the spot!

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx (hunk 5 above)
Comment (lhw-1) [reply src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx]: Addressed in [664d352](https://github.com/source-academy/frontend/pull/2836/commits/664d3528c4fc83f110bb07756f944cd7778ae46e).

File: src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx (hunk 6 above)
Comment (RichDom2185) [reply src/pages/academy/gameSimulator/subcomponents/chapterPublisher/ChapterPublisherEditor.tsx]: I don't mind using `{' '}` – I think it's used in a couple places as well.

I was just making sure since the change appeared out of nowhere; sure, we can leave it to the UI update, thanks!

**Review by gok99 — APPROVED** [review]
